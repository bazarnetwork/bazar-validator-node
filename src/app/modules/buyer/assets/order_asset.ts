import {
    ApplyAssetContext,
    BaseAsset,
    codec,
    StateStore,
    ValidateAssetContext } from 'lisk-sdk';
import { allOrdersSchema } from '../../seller/schema/order/allOrdersSchema';
import { orderSchema } from '../../seller/schema/order/orderSchema';
import { AllOrders } from '../../seller/types/order/allOrders';
import { OrderType } from '../../seller/types/order/orderType';
import { buyerOrderSchema } from '../schema/order/buyerOrderSchema';
import { registerBuyerOrderAssetSchema } from '../schema/order/registerBuyerOrderAsset';
import { RegisterBuyerOrderAccountType } from '../types/order/RegisterBuyerOrderAccountType';
import { RegisterOrderType } from '../types/order/registerBuyerOrderType';

const getAllOrders = async (stateStore: StateStore) => {
    const all = await stateStore.chain.get('purchasing_history/all');    
    if (all) {
        const orders: AllOrders = codec.decode(allOrdersSchema, all);
        return orders;
    } else {
        return {
            orders: [],
        };
    }
};

export class OrderAsset extends BaseAsset {
    public name = 'order';
    public id = 0;
    public schema = registerBuyerOrderAssetSchema;

    public validate({ asset }: ValidateAssetContext<RegisterOrderType>): void {
        if (asset.buyerOrderId.length <= 0) {
            throw new Error('Buyer Order Id is empty');
        } else if (asset.sellerOrderId.length <= 0) {
            throw new Error('Seller Order Id is empty');
		} else if (asset.status.length <= 0) {
            throw new Error('Status is empty');
        } else if (asset.token.length <= 0) {
            throw new Error('Token used in transaction is empty');
        } else if (asset.exchangeRate.length <= 0) {
            throw new Error('Exchange Rate is empty');
        } else if (asset.valueXKg.length <= 0) {
            throw new Error('Value X Kg is empty');
        } else if (asset.quantity <= 0) {
            throw new Error('Quantity hasnot value');
        } else if (asset.totalPayToken.length <= 0) {
            throw new Error('Total Payment in Token hasnot value');
        } else if (asset.totalPayInUSD.length <= 0) {
            throw new Error('Total Payment in USD hasnot value');
        } else if (asset.transacctionPayment.length <= 0) {
            throw new Error('Transacction Payment is empty');
        } else if (asset.accountSeller.length <= 0) {
			throw new Error('Account of seller is empty');
		} else if (asset.accountBuyer.length <= 0) {
			throw new Error('Account of buyer is empty');
		} else if (asset.productId.length <= 0) {
            throw new Error('Product Id is empty');
        }
    }

	// eslint-disable-next-line @typescript-eslint/require-await
    public async apply({ asset, transaction, stateStore }: ApplyAssetContext<RegisterOrderType>): Promise<void> {
	    const sender = await stateStore.account.get<RegisterBuyerOrderAccountType>(transaction.senderAddress);
        const orderBuffer = await stateStore.chain.get(asset.sellerOrderId);
        if (orderBuffer) {
            const decodedOrder = codec.decode<OrderType>(orderSchema, orderBuffer);

            if (decodedOrder.quantity >= asset.quantity) {

                const buyerOrder = {
                    id: asset.buyerOrderId,                    
                    sellerOrderId: asset.sellerOrderId,
                    status: asset.status,
                    token: asset.token,
                    exchangeRate: asset.exchangeRate,
                    valueXKg: asset.valueXKg,
                    quantity: asset.quantity,
                    serviceFee: asset.serviceFee,
                    totalPayToken: asset.totalPayToken,
                    totalPayInUSD: asset.totalPayInUSD,
                    transacctionPayment: asset.transacctionPayment,
                    accountSeller: asset.accountSeller,
                    accountBuyer: asset.accountBuyer,
                    productId: asset.productId,
                    date: Math.floor(Date.now() / 1000),
                    author: sender.address
                };

                await stateStore.chain.set(asset.buyerOrderId, codec.encode(buyerOrderSchema, buyerOrder));
                
                // updading inventory
                decodedOrder.quantity = decodedOrder.quantity - asset.quantity
                
                const allOrders: AllOrders = await getAllOrders(stateStore);
                allOrders.orders.push(asset.buyerOrderId);
                await stateStore.chain.set('purchasing_history/all', codec.encode(allOrdersSchema, allOrders));

                sender.buyer.orders.push(asset.buyerOrderId);
                await stateStore.chain.set(asset.sellerOrderId, codec.encode(orderSchema, decodedOrder));
                await stateStore.account.set(sender.address, sender);
            } else {
                throw new Error('Quantity to buy exceeds inventary');
            }
        }
	}
}
