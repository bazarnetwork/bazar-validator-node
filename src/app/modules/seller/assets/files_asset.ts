import { BaseAsset, ApplyAssetContext, ValidateAssetContext, cryptography, codec } from 'lisk-sdk';
import { fileRecordAssetSchema } from '../schema/files/fileRecordAsset';
import { orderSchema } from '../schema/order/orderSchema';
import { FileRecordType } from '../types/files/fileRecordType';
import { OrderType } from '../types/order/orderType';
import { RegisterOrderAccountType } from '../types/order/registerOrderAccountType';

const getId = (address: Buffer, nonce: bigint): Buffer => {
    const nonceBuffer = Buffer.alloc(8);
    nonceBuffer.writeBigInt64BE(nonce);
    const seed = Buffer.concat([address, nonceBuffer]);
    return cryptography.hash(seed);
};

export class FilesAsset extends BaseAsset {
	public name = 'files';
    public id = 1;
	public schema = fileRecordAssetSchema;

  public validate({ asset }: ValidateAssetContext<FileRecordType>): void {
    if (asset.filename.length <= 0) {
        throw new Error('Filename is empty');
    } else if (asset.fileType.length <= 0) {
        throw new Error('File Type is empty');
    } else if (asset.fileCategory.length <= 0) {
        throw new Error('File Category is empty');
    } else if (asset.hash.length <= 0) {
        throw new Error('Hash is empty');
    }
  }

	// eslint-disable-next-line @typescript-eslint/require-await
  public async apply({ asset, transaction, stateStore }: ApplyAssetContext<FileRecordType>): Promise<void> {
    const sender = await stateStore.account.get<RegisterOrderAccountType>(transaction.senderAddress);
    
    const orderBuffer = await stateStore.chain.get(asset.orderId);
    if (orderBuffer) {
        const decodedOrder = codec.decode<OrderType>(orderSchema, orderBuffer);
                
        const file = {
            filename: asset.filename,
            fileType: asset.fileType,
            fileCategory: asset.fileCategory,
            hash: asset.hash,
            date: Math.floor(Date.now() / 1000),
            author: sender.address
        };
        const fileId = getId(transaction.senderAddress, transaction.nonce).toString('hex');

        decodedOrder.files.push(file);

        sender.seller.files.push(`${asset.orderId}#${fileId}`);

        await stateStore.chain.set(asset.orderId, codec.encode(orderSchema, decodedOrder));
        await stateStore.account.set(sender.address, sender);
    }    
  } 
}
