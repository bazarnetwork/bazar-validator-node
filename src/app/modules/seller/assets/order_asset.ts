import {
    ApplyAssetContext,
    BaseAsset,
    codec,
    StateStore,
    ValidateAssetContext } from 'lisk-sdk';
import { allOrdersSchema } from '../schema/order/allOrdersSchema';    
import { AllOrders } from '../types/order/allOrders';
import { orderSchema } from '../schema/order/orderSchema';
import { registerOrderAssetSchema } from '../schema/order/registerOrderAsset';
import { RegisterOrderType } from '../types/order/registerOrderAssetType';
import { RegisterOrderAccountType } from '../types/order/registerOrderAccountType';

const getAllOrders = async (stateStore: StateStore) => {
    const all = await stateStore.chain.get('order/all');    
    if (all) {
        const orders: AllOrders = codec.decode(allOrdersSchema, all);
        return orders;
    } else {
        return {
            orders: [],
        };
    }
};

export class OrderAsset extends BaseAsset {	
	public name = 'order';
  	public id = 0;
	public schema = registerOrderAssetSchema;

	public validate({ asset }: ValidateAssetContext<RegisterOrderType>): void {
        if (asset.orderId.length <= 0) {
			throw new Error('Order Id is empty');
        } else if (asset.productId.length <= 0) {
			throw new Error('Product Id is empty');
		} else if (asset.productName.length <= 0) {
			throw new Error('Product Name is empty');
		} else if (asset.minQuantityToSell <= 0) {
			throw new Error('Min quantity to sell does not have a valid value');
		} else if (asset.quantity <= 0) {
			throw new Error('Quantity does not have a valid value');
		} else if (asset.price <= 0) {
			throw new Error('Price does not have a valid value');
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
  public async apply({ asset, transaction, stateStore }: ApplyAssetContext<RegisterOrderType>): Promise<void> {
		const sender = await stateStore.account.get<RegisterOrderAccountType>(transaction.senderAddress);

        let newfiles: any[] = [];

        if (asset.files) {
            newfiles = asset.files.map((file) => {
                return {
                    filename: file.filename,
                    fileType: file.fileType,
                    fileCategory: file.fileCategory,
                    hash: file.hash,
                    date: Math.floor(Date.now() / 1000),
                    author: sender.address
                };
            });
        }

        const order = {
            id: asset.orderId,
            productId: asset.productId,
            productName: asset.productName,
            productDescription: asset.productDescription,
            minQuantityToSell: asset.minQuantityToSell,
            quantity: asset.quantity,
            price: asset.price,
            files: newfiles,
            transport: [],
            date: Math.floor(Date.now() / 1000),
            author: sender.address
        };

        await stateStore.chain.set(asset.orderId, codec.encode(orderSchema, order));

        const allOrders: AllOrders = await getAllOrders(stateStore);
        allOrders.orders.push(asset.orderId);
        await stateStore.chain.set('order/all', codec.encode(allOrdersSchema, allOrders));

        sender.seller.orders.push(asset.orderId);
        await stateStore.account.set(sender.address, sender);
	}
}
