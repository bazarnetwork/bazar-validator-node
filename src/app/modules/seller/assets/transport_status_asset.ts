import { BaseAsset, ApplyAssetContext, ValidateAssetContext, cryptography, codec } from 'lisk-sdk';
import { orderSchema } from '../schema/order/orderSchema';
import { transportStatusAssetSchema } from '../schema/transportStatus/transportStatusAsset';
import { OrderType } from '../types/order/orderType';
import { RegisterOrderAccountType } from '../types/order/registerOrderAccountType';
import { TransportStatusType } from '../types/transportStatus/transportStatusType';

const getId = (address: Buffer, nonce: bigint): Buffer => {
    const nonceBuffer = Buffer.alloc(8);
    nonceBuffer.writeBigInt64BE(nonce);
    const seed = Buffer.concat([address, nonceBuffer]);
    return cryptography.hash(seed);
};

export class TransportStatusAsset extends BaseAsset {
	public name = 'transportStatus';
    public id = 2;

  // Define schema for asset
	public schema = transportStatusAssetSchema;

  public validate({ asset }: ValidateAssetContext<TransportStatusType>): void {
    if (asset.orderId.length <= 0) {
        throw new Error('Order Id is empty');
    } else if (asset.origin.length <= 0) {
        throw new Error('Origin is empty');
    } else if (asset.destiny.length <= 0) {
        throw new Error('Destiny is empty');
    } else if (asset.location.length <= 0) {
        throw new Error('Location is empty');
    } else if (asset.status.length <= 0) {
        throw new Error('Status is empty');
    }
  }

	// eslint-disable-next-line @typescript-eslint/require-await
  public async apply({ asset, transaction, stateStore }: ApplyAssetContext<TransportStatusType>): Promise<void> {
    const sender = await stateStore.account.get<RegisterOrderAccountType>(transaction.senderAddress);

    const orderBuffer = await stateStore.chain.get(asset.orderId);
    if (orderBuffer) {
        const decodedOrder = codec.decode<OrderType>(orderSchema, orderBuffer);

        const transport = {
            origin: asset.origin,
            destiny: asset.destiny,
            location: asset.location,
            date: Math.floor(Date.now() / 1000),
            status: asset.status,
            author: sender.address 
        };

        const transportStatusId = getId(transaction.senderAddress, transaction.nonce).toString('hex');

        decodedOrder.transport.push(transport);
        
        sender.seller.transport.push(`${asset.orderId}#${transportStatusId}`);

        await stateStore.chain.set(asset.orderId, codec.encode(orderSchema, decodedOrder));
        await stateStore.account.set(sender.address, sender);
    }
  }
}
