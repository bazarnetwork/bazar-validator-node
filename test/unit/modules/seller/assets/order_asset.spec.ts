import { OrderAsset } from '../../../../../src/app/modules/seller/assets/order_asset';

describe('OrderAsset', () => {
  let transactionAsset: OrderAsset;

	beforeEach(() => {
		transactionAsset = new OrderAsset();
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(0);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('order');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('validate', () => {
		describe('schema validation', () => {
      it.todo('should throw errors for invalid schema');
      it.todo('should be ok for valid schema');
    });
	});

	describe('apply', () => {
    describe('valid cases', () => {
      it.todo('should update the state store');
    });

    describe('invalid cases', () => {
      it.todo('should throw error');
    });
	});
});
