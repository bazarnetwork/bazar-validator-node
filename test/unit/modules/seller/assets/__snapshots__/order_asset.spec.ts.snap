// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderAsset constructor should have valid schema 1`] = `
Object {
  "$id": "bazar/seller/registerOrderAsset",
  "properties": Object {
    "minQuantityToSell": Object {
      "dataType": "uint64",
      "fieldNumber": 4,
    },
    "price": Object {
      "dataType": "uint64",
      "fieldNumber": 6,
    },
    "productDescription": Object {
      "dataType": "string",
      "fieldNumber": 3,
      "maxLength": 500,
    },
    "productId": Object {
      "dataType": "string",
      "fieldNumber": 1,
      "maxLength": 50,
    },
    "productName": Object {
      "dataType": "string",
      "fieldNumber": 2,
      "maxLength": 100,
    },
    "quantity": Object {
      "dataType": "uint64",
      "fieldNumber": 5,
    },
  },
  "required": Array [
    "productId",
    "productName",
    "productDescription",
    "quantity",
    "price",
  ],
  "title": "registerOrderAsset transaction asset for Seller module",
  "type": "object",
}
`;
